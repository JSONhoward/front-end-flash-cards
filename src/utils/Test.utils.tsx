import * as React from 'react'
import { RecoilRoot } from 'recoil'

export const RecoilWrapper: React.FC = ({children}) => {
    return (
        <RecoilRoot>
            {children}
        </RecoilRoot>
    )
}