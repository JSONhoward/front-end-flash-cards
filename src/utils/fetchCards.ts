export const fetchCards = async (): Promise<FirebaseDocArray> => {
    const response = await fetch(process.env.REACT_APP_GRAPHQL_URL as string, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'X-REQUEST-TYPE': 'GraphQL'
        },
        body: JSON.stringify({
            query: fetchAllCardsQuery
        })
    })
    const data = await response.json()
    const {data: {cards}} = data

    for (let i = cards.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [cards[i], cards[j]] = [cards[j], cards[i]]
    }

    return cards
}

const fetchAllCardsQuery = `
query {
    cards {
        id
        Category
        q
        code
        a
    }
}
`