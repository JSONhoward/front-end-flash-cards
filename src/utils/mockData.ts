export const mockQuestionArray: FirebaseDocArray = [
    {
        id: '1',
        Category: 'General',
        q: 'When building a new web site or maintaining one, can you explain some techniques you have used to increase performance?',
        a: 'some answer'
    },
    {
        id: '2',
        Category: 'General',
        code: 'some code',
        q: 'Test question 2',
        a: 'some answer'
    },
    {
        id: '3',
        Category: 'HTML',
        code: 'some code',
        q: 'Test question 3',
        a: 'some answer'
    },
    {
        id: '4',
        Category: 'Javascript',
        q: 'Test question 4',
        a: 'some answer'
    },
    {
        id: '5',
        Category: 'CSS',
        q: 'Test question 5',
        a: 'some answer'
    },
    {
        id: '6',
        Category: 'Networking',
        q: 'Test question 6',
        a: 'some answer'
    },
    {
        id: '7',
        Category: 'Testing',
        q: 'Test question 7',
        a: 'some answer'
    },
    {
        id: '8',
        Category: 'Coding',
        q: 'Test question 8',
        a: 'some answer'
    },
    {
        id: '9',
        Category: 'General',
        q: 'Test question 9',
        a: 'some answer'
    },
    {
        id: '10',
        Category: 'General',
        q: 'Test question 10',
        a: 'some answer'
    }
]