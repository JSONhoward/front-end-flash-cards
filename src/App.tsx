import React from 'react';
import { useSetRecoilState } from 'recoil';
import Home from './pages/Home';
import { AllQuestions } from './store/atoms';
import { mockQuestionArray } from './utils/mockData';
import { FirebaseContext } from './Firebase/index';

type Props = {
  fetchCards: (firebase: firebase.app.App) => Promise<FirebaseDocArray>
}

const App: React.FC<Props> = ({fetchCards}) => {
  const firebase = React.useContext<firebase.app.App | null>(FirebaseContext)
  const fetchAllQuestions = useSetRecoilState(AllQuestions)

  React.useEffect(() => {
    fetchCards(firebase as firebase.app.App)
      .then(cards => fetchAllQuestions(cards))
    // fetchAllQuestions(mockQuestionArray)
    // eslint-disable-next-line
  }, [])

  return (
    <Home />
  )
}

export default App;
