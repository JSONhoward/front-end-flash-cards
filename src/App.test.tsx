import React from 'react'
import {unmountComponentAtNode, render} from 'react-dom'

import App from './App'
import { RecoilWrapper } from './utils/Test.utils'

const fetchCards = jest.fn((firebase: any): Promise<FirebaseDocArray> => {
    return new Promise((res, _) => {
        res(
            [
                {
                    id: '1',
                    Category: 'test category',
                    q: 'question 1',
                    code: 'some code',
                    a: 'answer'
                },
                {
                    id: '2',
                    Category: 'test category',
                    q: 'question 2',
                    a: 'answer'
                }
            ]
        )
    })
})

let container: HTMLDivElement | null = null
beforeEach(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
})

afterEach(() => {
    unmountComponentAtNode(container as HTMLDivElement)
    container?.remove()
    container = null
})

describe('App Tests', () => {
    it('Matches Snapshot', () => {
        render(<RecoilWrapper><App fetchCards={fetchCards} /></RecoilWrapper>, container)

        expect(container?.innerHTML).toMatchSnapshot()
    })

    it('Calls firebase api', () => {
        render(<RecoilWrapper><App fetchCards={fetchCards} /></RecoilWrapper>, container)

        expect(fetchCards).toHaveBeenCalledTimes(1)
    })
})