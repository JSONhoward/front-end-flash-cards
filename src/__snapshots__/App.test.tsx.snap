// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Tests Matches Snapshot 1`] = `"<div class=\\"sc-jrsJCI gzNwVG\\"><h1 style=\\"color: whitesmoke; text-align: center;\\">Front End Developer Interview Questions</h1><div class=\\"sc-hKFyIo eIJgGR\\"><div class=\\"sc-gKAblj PdYxk\\"><div offset=\\"0\\" class=\\"sc-dlnjPT cqeDqx\\"><div class=\\"sc-bdnylx iZSjZb\\">Loading...</div></div></div><p class=\\"sc-jSFkmK dCqQGP\\">Question# 1 of 1</p><div class=\\"sc-eCApGN fGzqiL\\"><div class=\\"sc-iCoHVE enmkFP\\"><label for=\\"category\\">Category:</label><select name=\\"category\\"><option value=\\"all\\">All</option><option value=\\"General\\">General</option><option value=\\"HTML\\">HTML</option><option value=\\"CSS\\">CSS</option><option value=\\"Javascript\\">Javascript</option><option value=\\"Networking\\">Networking</option><option value=\\"Testing\\">Testing</option></select></div><div data-testid=\\"showAnswer\\" class=\\"sc-pNWxx lcrklA\\"><p>Show Answer</p></div><div data-testid=\\"nextCard\\" class=\\"sc-fujyUd faXeHE\\">Next Card</div></div></div></div>"`;
